dol/sources.py

```diff
@@ -150,6 +227,177 @@ def __contains__(self, k) -> int:
+
+    Note that the value of the non-existing store key is ignored! So, be careful when
+    using this option, to avoid losing data.
+
+    Let's delete items now:
+
+    >>> del persister['a']
+    >>> 'a' in persister
+    False
+
+    The key as been deleted from all the stores:
+
+    >>> 'a' in bytes_store
+    False
+    >>> 'a' in metadata_store
+    False
+
+    As expected, if the key is not in any of the stores, a KeyError is raised:
+
+    >>> del persister['z']
+    Traceback (most recent call last):
+        ...
+    KeyError: 'z'
+
+    However, if the key is in some of the stores, but not in others, a KeyError is raised
+    only if `need_to_set_all_stores=True`:
+
+    >>> persister = FanoutPersister(
+    ...     bytes_store=dict(a=b'a'),
+    ...     metadata_store=dict(),
+    ...     need_to_set_all_stores=True,
+    ... )
+    >>> del persister['a']
+    Traceback (most recent call last):
+        ...
+    KeyError: "The key 'a' is not present in the following stores: {'metadata_store'}"
+
+    Otherwise, the key is deleted from the stores where it is present:
+
+    >>> bytes_store=dict(a=b'a')
+    >>> persister = FanoutPersister(
+    ...     bytes_store=bytes_store,
+    ...     metadata_store=dict(),
+    ... )
+    >>> del persister['a']
+    >>> 'a' in persister
+    False
+    >>> 'a' in bytes_store
+    False
+    '''
+    def __init__(
+            self,
+            *args,
+            default=None,
+            get_existing_values_only=False,
+            need_to_set_all_stores=False,
+            ignore_non_existing_store_keys=False,
+            **kwargs
+        ):
+        super().__init__(
+            *args,
+            default=default,
+            get_existing_values_only=get_existing_values_only,
+            **kwargs,
+        )
+        self._need_to_set_all_stores = need_to_set_all_stores
+        self._ignore_non_existing_store_keys = ignore_non_existing_store_keys
+
+    def __setitem__(self, k, v: Mapping):
+        if self._need_to_set_all_stores and not set(self._stores).issubset(set(v)):
+            missing_stores = set(self._stores) - set(v)
+            raise ValueError(
+                f'All stores must be set when setting a value. Missing stores: {missing_stores}'
+            )
+        if not self._ignore_non_existing_store_keys and not set(v).issubset(set(self._stores)):
+            invalid_store_keys = set(v) - set(self._stores)
+            raise ValueError(f'The value contains some invalid store keys: {invalid_store_keys}')
+        for store_key, vv in v.items():
+            if store_key in self._stores:
+                self._stores[store_key][k] = vv
+
+    def __delitem__(self, k):
+        stores_to_delete_from = {
+            store_key: store for store_key, store in self._stores.items() if k in store
+        }
+        if not stores_to_delete_from:
+            raise KeyError(k)
+        if self._need_to_set_all_stores and len(stores_to_delete_from) < len(self._stores):
```

Comment:
Perhaps there's a more general term for `_need_to_set_all_stores`, since we're really saying that all keys need to be present when we set, but also delete, and possibly read (should we check on this when doing a `getitem`?) etc.

Not a biggy, but just mentioning.

Reply:
The last comment is also one that points out the inconsistency. 
It's fine to require that all stores be written to on set (`_need_to_set_all_stores`), and you're carrying on that condition (which is expressed only as a condition on setitem) to deletion... but not on getitem?

It's a choice. We don't want to HAVE to burden getitem with these checks, that may be overkill (since we enforce it on the setitem) -- But then why do we enforce it systematically (if `_need_to_set_all_stores` is `True`) on deletion?

Reply:
True, _need_to_set_all_stores should only apply to set. And let's getitem and delitem permissive